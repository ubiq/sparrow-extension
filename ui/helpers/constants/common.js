export const ETH = 'UBQ';
export const GWEI = 'GWEI';
export const WEI = 'WEI';

export const PRIMARY = 'PRIMARY';
export const SECONDARY = 'SECONDARY';

export const ERC20 = 'ERC20';
export const ERC721 = 'ERC721';
export const ERC1155 = 'ERC1155';

/**
 * @typedef {Object} TokenStandards
 * @property {'ERC20'} ERC20 - A token that conforms to the ERC20 standard.
 * @property {'ERC721'} ERC721 - A token that conforms to the ERC721 standard.
 * @property {'ERC1155'} ERC1155 - A token that conforms to the ERC1155
 *  standard.
 * @property {'NONE'} NONE - Not a token, but rather the base asset of the
 *  selected chain.
 */

/**
 * This type will work anywhere you expect a string that can be one of the
 * above statuses
 *
 * @typedef {TokenStandards[keyof TokenStandards]} TokenStandardStrings
 */

/**
 * Describes the standard which a token conforms to.
 *
 * @type {TokenStandards}
 */
export const TOKEN_STANDARDS = {
  ERC20,
  ERC721,
  ERC1155,
  NONE: 'NONE',
};

export const GAS_ESTIMATE_TYPES = {
  SLOW: 'SLOW',
  AVERAGE: 'AVERAGE',
  FAST: 'FAST',
  FASTEST: 'FASTEST',
};

let _supportLink = 'https://support.metamask.io';
let _supportRequestLink = 'https://metamask.zendesk.com/hc/en-us/requests/new';
const _contractAddressLink =
  'https://metamask.zendesk.com/hc/en-us/articles/360020028092-What-is-the-known-contract-address-warning-';

///: BEGIN:ONLY_INCLUDE_IN(flask)
_supportLink = 'https://metamask-flask.zendesk.com/hc';
_supportRequestLink =
  'https://metamask-flask.zendesk.com/hc/en-us/requests/new';
///: END:ONLY_INCLUDE_IN

export const SUPPORT_LINK = _supportLink;
export const SUPPORT_REQUEST_LINK = _supportRequestLink;
export const CONTRACT_ADDRESS_LINK = _contractAddressLink;
