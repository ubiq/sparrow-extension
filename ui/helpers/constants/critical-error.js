export const TEN_SECONDS_IN_MILLISECONDS = 10_000;
