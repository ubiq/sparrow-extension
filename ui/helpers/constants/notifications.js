///: BEGIN:ONLY_INCLUDE_IN(flask)

// The time after which a notification should be deleted.
export const NOTIFICATIONS_EXPIRATION_DELAY = 10000;

///: END:ONLY_INCLUDE_IN
