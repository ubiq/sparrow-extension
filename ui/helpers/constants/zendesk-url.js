const ZENDESK_URLS = {
  ADD_CUSTOM_TOKENS:
    'https://metamask.zendesk.com/hc/en-us/articles/360015489031',
  ADD_MISSING_ACCOUNTS:
    'https://metamask.zendesk.com/hc/en-us/articles/360015489271',
  IMPORT_ACCOUNTS:
    'https://metamask.zendesk.com/hc/en-us/articles/360015489331',
  TOKEN_SAFETY_PRACTICES:
    'https://metamask.zendesk.com/hc/en-us/articles/4403988839451',
  SECRET_RECOVERY_PHRASE:
    'https://metamask.zendesk.com/hc/en-us/articles/360060826432-What-is-a-Secret-Recovery-Phrase-and-how-to-keep-your-crypto-wallet-secure',
  PASSWORD_ARTICLE:
    'https://metamask.zendesk.com/hc/en-us/articles/4404722782107',
};

export default ZENDESK_URLS;
