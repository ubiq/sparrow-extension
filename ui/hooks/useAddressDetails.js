import { useSelector } from 'react-redux';

import { toChecksumHexAddress } from '../../shared/modules/hexstring-utils';
import {
  getAddressBook,
  getMetaMaskIdentities,
  getTokenList,
  getUseTokenDetection,
} from '../selectors';
import { shortenAddress } from '../helpers/utils/util';

const useAddressDetails = (toAddress) => {
  const addressBook = useSelector(getAddressBook);
  const identities = useSelector(getMetaMaskIdentities);
  const tokenList = useSelector(getTokenList);
  const useTokenDetection = useSelector(getUseTokenDetection);
  const checksummedAddress = toChecksumHexAddress(toAddress);

  if (!toAddress) {
    return {};
  }
  const addressBookEntryObject = addressBook.find(
    (entry) => entry.address === checksummedAddress,
  );
  if (addressBookEntryObject?.name) {
    return { toName: addressBookEntryObject.name, isTrusted: true };
  }
  if (identities[toAddress]?.name) {
    return { toName: identities[toAddress].name, isTrusted: true };
  }
  if (process.env.TOKEN_DETECTION_V2) {
    if (tokenList[toAddress]?.name) {
      return { toName: tokenList[toAddress].name, isTrusted: true };
    }
  } else {
    const casedTokenList = useTokenDetection
      ? tokenList
      : Object.keys(tokenList).reduce((acc, base) => {
          return {
            ...acc,
            [base.toLowerCase()]: tokenList[base],
          };
        }, {});
    if (casedTokenList[toAddress]?.name) {
      return { toName: casedTokenList[toAddress].name, isTrusted: true };
    }
  }
  return {
    toName: shortenAddress(checksummedAddress),
    isTrusted: false,
  };
};

export default useAddressDetails;
