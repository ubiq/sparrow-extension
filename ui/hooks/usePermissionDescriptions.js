import { useMemo } from 'react';
import {
  RestrictedMethods,
  ///: BEGIN:ONLY_INCLUDE_IN(flask)
  EndowmentPermissions,
  PermissionNamespaces,
  ///: END:ONLY_INCLUDE_IN
} from '../../shared/constants/permissions';
///: BEGIN:ONLY_INCLUDE_IN(flask)
import { coinTypeToProtocolName } from '../helpers/utils/util';
///: END:ONLY_INCLUDE_IN
import { useI18nContext } from './useI18nContext';

const UNKNOWN_PERMISSION = Symbol('unknown');

/**
 * @typedef {Object} PermissionLabelObject
 * @property {string} label - The text label.
 * @property {string} leftIcon - The left icon.
 * @property {string} rightIcon - The right icon.
 */

/**
 * @returns {(permissionName:string) => PermissionLabelObject}
 */
export const usePermissionDescriptions = () => {
  const t = useI18nContext();

  return useMemo(() => {
    const permissionDescriptions = {
      [RestrictedMethods.eth_accounts]: {
        leftIcon: 'fas fa-eye',
        label: t('permission_ethereumAccounts'),
        rightIcon: null,
      },
      ///: BEGIN:ONLY_INCLUDE_IN(flask)
      [RestrictedMethods.snap_confirm]: {
        leftIcon: 'fas fa-user-check',
        label: t('permission_customConfirmation'),
        rightIcon: null,
      },
      [RestrictedMethods['snap_getBip44Entropy_*']]: (permissionName) => {
        const coinType = permissionName.split('_').slice(-1);
        return {
          leftIcon: 'fas fa-door-open',
          label: t('permission_manageBip44Keys', [
            coinTypeToProtocolName(coinType) ||
              `${coinType} (Unrecognized protocol)`,
          ]),
          rightIcon: null,
        };
      },
      [RestrictedMethods.snap_manageState]: {
        leftIcon: 'fas fa-download',
        label: t('permission_manageState'),
        rightIcon: null,
      },
      [RestrictedMethods['wallet_snap_*']]: (permissionName) => {
        const snapId = permissionName.split('_').slice(-1);
        return {
          leftIcon: 'fas fa-bolt',
          label: t('permission_accessSnap', [snapId]),
          rightIcon: null,
        };
      },
      [EndowmentPermissions['endowment:network-access']]: {
        leftIcon: 'fas fa-wifi',
        label: t('permission_accessNetwork'),
        rightIcon: null,
      },
      ///: END:ONLY_INCLUDE_IN
      [UNKNOWN_PERMISSION]: (permissionName) => {
        return {
          leftIcon: 'fas fa-times-circle',
          label: t('permission_unknown', [permissionName ?? 'undefined']),
          rightIcon: null,
        };
      },
    };

    return (permissionName) => {
      let value = permissionDescriptions[UNKNOWN_PERMISSION];

      if (Object.hasOwnProperty.call(permissionDescriptions, permissionName)) {
        value = permissionDescriptions[permissionName];
      }
      ///: BEGIN:ONLY_INCLUDE_IN(flask)
      for (const namespace of Object.keys(PermissionNamespaces)) {
        if (permissionName.startsWith(namespace)) {
          value = permissionDescriptions[PermissionNamespaces[namespace]];
        }
      }
      ///: END:ONLY_INCLUDE_IN

      return typeof value === 'function' ? value(permissionName) : value;
    };
  }, [t]);
};
