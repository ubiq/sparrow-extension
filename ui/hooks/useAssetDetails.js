import { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getCollectibles, getTokens } from '../ducks/metamask/metamask';
import { ERC1155, ERC721, ERC20 } from '../helpers/constants/common';
import {
  calcTokenAmount,
  getAssetDetails,
  getTokenAddressParam,
  getTokenValueParam,
} from '../helpers/utils/token-util';
import { getTransactionData } from '../helpers/utils/transactions.util';
import { getTokenList } from '../selectors';
import { hideLoadingIndication, showLoadingIndication } from '../store/actions';
import { usePrevious } from './usePrevious';

export function useAssetDetails(tokenAddress, userAddress, transactionData) {
  const dispatch = useDispatch();

  // state selectors
  const tokens = useSelector(getTokens);
  const collectibles = useSelector(getCollectibles);
  const tokenList = useSelector(getTokenList);

  // in-hook state
  const [currentAsset, setCurrentAsset] = useState(null);

  // previous state checkers
  const prevTokenAddress = usePrevious(tokenAddress);
  const prevUserAddress = usePrevious(userAddress);
  const prevTransactionData = usePrevious(transactionData);

  useEffect(() => {
    async function getAndSetAssetDetails() {
      dispatch(showLoadingIndication());
      const assetDetails = await getAssetDetails(
        tokenAddress,
        userAddress,
        transactionData,
        collectibles,
        tokens,
        tokenList,
      );
      setCurrentAsset(assetDetails);
      dispatch(hideLoadingIndication());
    }
    if (
      tokenAddress !== prevTokenAddress ||
      userAddress !== prevUserAddress ||
      transactionData !== prevTransactionData
    ) {
      getAndSetAssetDetails();
    }
  }, [
    dispatch,
    prevTokenAddress,
    prevTransactionData,
    prevUserAddress,
    tokenAddress,
    userAddress,
    transactionData,
    collectibles,
    tokens,
    tokenList,
  ]);

  let assetStandard,
    assetName,
    assetAddress,
    tokenSymbol,
    decimals,
    tokenImage,
    userBalance,
    tokenValue,
    toAddress,
    tokenAmount,
    tokenId;

  if (currentAsset) {
    const {
      standard,
      symbol,
      image,
      name,
      balance,
      decimals: currentAssetDecimals,
    } = currentAsset;
    const tokenData = getTransactionData(transactionData);
    assetStandard = standard;
    assetAddress = tokenAddress;
    tokenSymbol = symbol;
    tokenImage = image;
    toAddress = getTokenAddressParam(tokenData);
    if (assetStandard === ERC721 || assetStandard === ERC1155) {
      assetName = name;
      tokenId = getTokenValueParam(tokenData);
    }
    if (assetStandard === ERC20) {
      userBalance = balance;
      decimals = Number(currentAssetDecimals?.toString(10));
      tokenAmount =
        tokenData &&
        calcTokenAmount(getTokenValueParam(tokenData), decimals).toString(10);
    }
  }
  return {
    assetStandard,
    assetName,
    assetAddress,
    userBalance,
    tokenSymbol,
    decimals,
    tokenImage,
    tokenValue,
    toAddress,
    tokenAmount,
    tokenId,
  };
}
