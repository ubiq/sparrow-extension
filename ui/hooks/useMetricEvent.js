import { useContext, useCallback } from 'react';
import { useSelector } from 'react-redux';
import { useRouteMatch } from 'react-router-dom';
import { MetaMetricsContext } from '../contexts/metametrics';
import { MetaMetricsContext as NewMetaMetricsContext } from '../contexts/metametrics.new';
import { PATH_NAME_MAP } from '../helpers/constants/routes';
import { txDataSelector } from '../selectors';
import { useEqualityCheck } from './useEqualityCheck';

// Type imports
/**
 * @typedef {import('../contexts/metametrics.new').UIMetricsEventPayload} UIMetricsEventPayload
 * @typedef {import('../../shared/constants/metametrics').MetaMetricsEventOptions} MetaMetricsEventOptions
 */

export function useMetricEvent(config = {}, overrides = {}) {
  const metricsEvent = useContext(MetaMetricsContext);
  const trackEvent = useCallback(() => metricsEvent(config, overrides), [
    config,
    metricsEvent,
    overrides,
  ]);
  return trackEvent;
}

/**
 * track a metametrics event using segment
 * e.g metricsEvent({ event: 'Unlocked Sparrow', category: 'Navigation' })
 *
 * @param {UIMetricsEventPayload} payload - payload of the event to track
 * @param {MetaMetricsEventOptions} options - options for handling/routing event
 * @returns {() => Promise<void>} function to execute the tracking event
 */
export function useNewMetricEvent(payload, options) {
  const memoizedPayload = useEqualityCheck(payload);
  const memoizedOptions = useEqualityCheck(options);
  const metricsEvent = useContext(NewMetaMetricsContext);

  return useCallback(() => metricsEvent(memoizedPayload, memoizedOptions), [
    metricsEvent,
    memoizedPayload,
    memoizedOptions,
  ]);
}

const PATHS_TO_CHECK = Object.keys(PATH_NAME_MAP);

/**
 * Returns the current page if it matches our route map, as well as the origin
 * if there is a confirmation that was triggered by a dapp. These values are
 * not required but add valuable context to events, and should be included in
 * the context object on the event payload.
 *
 * @returns {{
 *  page?: MetaMetricsPageObject
 *  referrer?: MetaMetricsReferrerObject
 * }}
 */
export function useMetaMetricsContext() {
  const match = useRouteMatch({
    path: PATHS_TO_CHECK,
    exact: true,
    strict: true,
  });
  const txData = useSelector(txDataSelector) || {};
  const confirmTransactionOrigin = txData.origin;

  const referrer = confirmTransactionOrigin
    ? {
        url: confirmTransactionOrigin,
      }
    : undefined;

  const page = match
    ? {
        path: match.path,
        title: PATH_NAME_MAP[match.path],
        url: match.path,
      }
    : undefined;

  return {
    page,
    referrer,
  };
}
