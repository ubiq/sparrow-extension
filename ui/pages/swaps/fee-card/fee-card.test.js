import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { useSelector } from 'react-redux';

import {
  renderWithProvider,
  createSwapsMockStore,
  setBackgroundConnection,
  MOCKS,
} from '../../../../test/jest';
import { EDIT_GAS_MODES } from '../../../../shared/constants/gas';
import { MAINNET_CHAIN_ID } from '../../../../shared/constants/network';

import {
  checkNetworkAndAccountSupports1559,
  getEIP1559V2Enabled,
  getPreferences,
  getSelectedAccount,
} from '../../../selectors';
import {
  getGasEstimateType,
  getGasFeeEstimates,
  getIsGasEstimatesLoading,
} from '../../../ducks/metamask/metamask';
import { GasFeeContextProvider } from '../../../contexts/gasFee';
import { TRANSACTION_ENVELOPE_TYPE_NAMES } from '../../../helpers/constants/transactions';
import { useGasFeeEstimates } from '../../../hooks/useGasFeeEstimates';

import FeeCard from '.';

const middleware = [thunk];

jest.mock('../../../hooks/useGasFeeEstimates', () => ({
  useGasFeeEstimates: jest.fn(),
}));

jest.mock('react-redux', () => {
  const actual = jest.requireActual('react-redux');

  return {
    ...actual,
    useSelector: jest.fn(),
  };
});

const generateUseSelectorRouter = () => (selector) => {
  if (selector === checkNetworkAndAccountSupports1559) {
    return true;
  }
  if (selector === getGasEstimateType) {
    return TRANSACTION_ENVELOPE_TYPE_NAMES.FEE_MARKET;
  }
  if (selector === getGasFeeEstimates) {
    return MOCKS.createGasFeeEstimatesForFeeMarket();
  }
  if (selector === getIsGasEstimatesLoading) {
    return false;
  }
  return undefined;
};

setBackgroundConnection({
  getGasFeeTimeEstimate: jest.fn(),
  getGasFeeEstimatesAndStartPolling: jest.fn(),
  createTransactionEventFragment: jest.fn(),
});

const createProps = (customProps = {}) => {
  return {
    primaryFee: {
      fee: '0.0441 ETH',
      maxFee: '0.04851 ETH',
    },
    secondaryFee: {
      fee: '$101.98',
      maxFee: '$112.17',
    },
    hideTokenApprovalRow: false,
    onFeeCardMaxRowClick: jest.fn(),
    tokenApprovalTextComponent: (
      <span key="fee-card-approve-symbol" className="fee-card__bold">
        ABC
      </span>
    ),
    tokenApprovalSourceTokenSymbol: 'ABC',
    onTokenApprovalClick: jest.fn(),
    metaMaskFee: '0.875',
    isBestQuote: true,
    numberOfQuotes: 6,
    onQuotesClick: jest.fn(),
    tokenConversionRate: 0.015,
    chainId: MAINNET_CHAIN_ID,
    networkAndAccountSupports1559: false,
    supportsEIP1559V2: false,
    ...customProps,
  };
};

describe('FeeCard', () => {
  it('renders the component with initial props', () => {
    useSelector.mockImplementation(generateUseSelectorRouter());
    const props = createProps();
    const { getByText } = renderWithProvider(<FeeCard {...props} />);
    expect(getByText('Best of 6 quotes.')).toBeInTheDocument();
    expect(getByText('Estimated gas fee')).toBeInTheDocument();
    expect(getByText('Max fee')).toBeInTheDocument();
    expect(getByText(props.primaryFee.fee)).toBeInTheDocument();
    expect(getByText(props.secondaryFee.fee)).toBeInTheDocument();
    expect(getByText(`: ${props.secondaryFee.maxFee}`)).toBeInTheDocument();
    expect(getByText('Includes a 0.875% Sparrow fee.')).toBeInTheDocument();
    expect(
      document.querySelector('.fee-card__top-bordered-row'),
    ).toMatchSnapshot();
  });

  it('renders the component with EIP-1559 enabled', () => {
    const store = configureMockStore(middleware)(createSwapsMockStore());
    const props = createProps({
      networkAndAccountSupports1559: true,
      maxPriorityFeePerGasDecGWEI: '3',
      maxFeePerGasDecGWEI: '4',
    });
    const { getByText } = renderWithProvider(<FeeCard {...props} />, store);
    expect(getByText('Best of 6 quotes.')).toBeInTheDocument();
    expect(getByText('Estimated gas fee')).toBeInTheDocument();
    expect(getByText('Max fee')).toBeInTheDocument();
    expect(getByText(props.primaryFee.fee)).toBeInTheDocument();
    expect(getByText(props.secondaryFee.fee)).toBeInTheDocument();
    expect(getByText(`: ${props.secondaryFee.maxFee}`)).toBeInTheDocument();
    expect(getByText('Includes a 0.875% Sparrow fee.')).toBeInTheDocument();
    expect(
      document.querySelector('.fee-card__top-bordered-row'),
    ).toMatchSnapshot();
  });

  it('renders the component with EIP-1559 V2 enabled', () => {
    useGasFeeEstimates.mockImplementation(() => ({ gasFeeEstimates: {} }));
    useSelector.mockImplementation((selector) => {
      if (selector === getPreferences) {
        return {
          useNativeCurrencyAsPrimaryCurrency: true,
        };
      }
      if (selector === getEIP1559V2Enabled) {
        return true;
      }
      if (selector === getSelectedAccount) {
        return {
          balance: '0x440aa47cc2556',
        };
      }
      if (selector === checkNetworkAndAccountSupports1559) {
        return true;
      }
      return undefined;
    });

    const store = configureMockStore(middleware)(createSwapsMockStore());
    const props = createProps({
      networkAndAccountSupports1559: true,
      maxPriorityFeePerGasDecGWEI: '3',
      maxFeePerGasDecGWEI: '4',
      supportsEIP1559V2: true,
    });
    const { getByText } = renderWithProvider(
      <GasFeeContextProvider
        transaction={{ txParams: {}, userFeeLevel: 'high' }}
        editGasMode={EDIT_GAS_MODES.SWAPS}
      >
        <FeeCard {...props} />
      </GasFeeContextProvider>,
      store,
    );
    expect(getByText('Best of 6 quotes.')).toBeInTheDocument();
    expect(getByText('Gas')).toBeInTheDocument();
    expect(getByText('(estimated)')).toBeInTheDocument();
    expect(getByText('Swap suggested')).toBeInTheDocument();
    expect(getByText('Max fee')).toBeInTheDocument();
    expect(getByText(props.primaryFee.fee)).toBeInTheDocument();
    expect(getByText(props.secondaryFee.fee)).toBeInTheDocument();
    expect(getByText(`: ${props.secondaryFee.maxFee}`)).toBeInTheDocument();
    expect(getByText('Includes a 0.875% Sparrow fee.')).toBeInTheDocument();
    expect(
      document.querySelector('.fee-card__top-bordered-row'),
    ).toMatchSnapshot();
  });
});
