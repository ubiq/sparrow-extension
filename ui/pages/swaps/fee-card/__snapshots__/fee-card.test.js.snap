// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeeCard renders the component with EIP-1559 enabled 1`] = `null`;

exports[`FeeCard renders the component with initial props 1`] = `null`;
